core/src/main/scala/org/ensime/core/Analyzer.scala

```diff
@@ -185,18 +185,29 @@ class Analyzer(
       sender ! VoidResponse
     case TypecheckModule(moduleName) =>
       //consider the case of a project with no modules
-      config.modules get (moduleName) foreach {
-        module =>
-          val files: Set[SourceFileInfo] = module.scalaSourceFiles.map(s => SourceFileInfo(EnsimeFile(s), None, None))(breakOut)
-          sender ! scalaCompiler.handleReloadFiles(files)
+      val responses: List[RpcResponse] = for (module <- (config.modules filterKeys (_.project == moduleName)).values.toList) yield {
+        val files: Set[SourceFileInfo] = module.scalaSourceFiles.map(s => SourceFileInfo(EnsimeFile(s), None, None))(breakOut)
+        scalaCompiler.handleReloadFiles(files)
+      }
+      val (_, errors) = responses.partition(_ == VoidResponse)
+      if (errors.isEmpty)
+        sender ! VoidResponse
+      else {
+        val accumulatedErrorString = (errors foldLeft "") { (allErrors, err) =>
+          err match {
+            case EnsimeServerError(desc) => allErrors + "\n" + desc
+            case _ => allErrors
+          }
+        }.tail
+        sender ! EnsimeServerError(accumulatedErrorString)
       }
     case UnloadModuleReq(moduleName) =>
-      config.modules get (moduleName) foreach {
+      (config.modules filterKeys (_.project == moduleName)).values foreach {
         module =>
           val files = module.scalaSourceFiles.toList
           files.foreach(scalaCompiler.askRemoveDeleted)
-          sender ! VoidResponse
       }
+      sender ! VoidResponse
```

Comment:
But for `TypeCheckModule` we maybe better off, accepting an `id`, it would be much simpler to change the call site to send the configuration.